internal/processor/filtermetric/config.go

```diff
@@ -16,13 +16,31 @@ package filtermetric
 
 import (
 	"go.opentelemetry.io/collector/internal/processor/filterset"
+	"go.opentelemetry.io/collector/internal/processor/filterset/regexp"
+)
+
+// MatchType specifies the strategy for matching against `pdata.Metric`s. This
+// is distinct from filterset.MatchType which matches against metric (and
+// tracing) names only. To support matching against metric names and
+// `pdata.Metric`s, filtermetric.MatchType is effectively a superset of
+// filterset.MatchType.
+type MatchType string
```

Comment:
Is there any possibility not to introduce a new type for this? Are we adding this new functionality for metrics only? It appears it could be useful for spans and logs as well. Can we make it uniform?

Reply:
I initially thought we could move `MatchType` to the `filterconfig` package and put all three match types there. But the code in `filterconfig` is currently just for filtering spans, and it's not used by `filterprocessor` so it didn't seem like `MatchType` belonged. In addition, moving `MatchType` to the `filterconfig` package would cause an import cycle.

However, what do you think about moving the contents of `filterconfig` to a new package called `filterspanconfig` (since that code is only used for filtering spans)? Then `MatchType` could be put in `filterconfig`. Maybe this would be a good home for it especially if we eventually want to also use expr filtering for traces/spans/logs?

If this sounds okay I can create a PR to rename `filterconfig` to `filterspanconfig`.

Reply:
There is some log filtering logic in `filterconfig` as well, so a simple rename to `filterspanconfig` isn't going to work. I'll look into it.

Reply:
So, to recap, I don't think we want to add the `expr` MatchType to the `filterset.MatchType` constants because `expr` doesn't apply to `filterset`. We could move `filterset.MatchType` to the `filterconfig` package and then add `expr`, but to prevent a dependency cycle `filterset.Config` would also have to move. Doable, but there's actually a fair number of references that would also get updated.

Because the `filterprocessor` doesn't filter anything but metrics, what do you think about keeping the solution in this PR and then doing any consolidation necessary if/when `expr` becomes an option for spans and logs? We may well want to simply revamp how we configure filters before then anyway.

Reply:
Another option is to move `filtermetric.MatchType` to the `filterconfig` package. That's easy to do, but we'll still have the additional type.

Reply:
@pmcollins Sounds good, thanks for the investigation.